### PR TITLE
Fix the Division by Zero Bug of CosineAnnealingLR

### LINKED ISSUE #17913: Bug in CosineAnnealingLR (division by zero)
## 🐛 Bug

The `CosineAnnealingLR` throws a `ZeroDivisionError` after being applied for more than one epoch.
Also, the learning rates returned by the `CosineAnnealingLR` do not match the internal learning rates of the passed `optimizer`.

## To Reproduce
```python
import torch
import torch.nn as nn
import torch.optim as optim
from torch.optim.lr_scheduler import _LRScheduler
import math

# Test new scheduler
model = nn.Linear(10, 2)
optimizer = optim.SGD(model.parameters(), lr=1.)
steps = 10
scheduler = optim.lr_scheduler.CosineAnnealingLR(optimizer, steps)

try:
    for epoch in range(2):
        for idx in range(steps):
            scheduler.step()
            print(scheduler.get_lr())
            print(optimizer.param_groups[0]['lr'])
except ZeroDivisionError as e:
    print(e)

> [1.0]
1.0
[0.9516553824444453]
0.9755282581475768
[0.8386590697412432]
0.9045084971874737
[0.6968044012954536]
0.7938926261462366
[0.5395961100811341]
0.6545084971874737
[0.38196601125010515]
0.5
[0.23872875703131577]
0.34549150281252633
[0.12295598939793909]
0.2061073738537635
[0.04424211972088169]
0.09549150281252633
[0.006271407734229157]
0.024471741852423234
[0.0]
0.0
float division by zero
```

## Expected behavior

Using the legacy implementation, no exception is raised and the learning rates of the scheduler and optimizer are equal:
```python
# Compare with legacy implementation
class LegacyCosineAnnealingLR(_LRScheduler):
    r"""Set the learning rate of each parameter group using a cosine annealing
    schedule, where :math:`\eta_{max}` is set to the initial lr and
    :math:`T_{cur}` is the number of epochs since the last restart in SGDR:

    .. math::

        \eta_t = \eta_{min} + \frac{1}{2}(\eta_{max} - \eta_{min})(1 +
        \cos(\frac{T_{cur}}{T_{max}}\pi))

    When last_epoch=-1, sets initial lr as lr.

    It has been proposed in
    `SGDR: Stochastic Gradient Descent with Warm Restarts`_. Note that this only
    implements the cosine annealing part of SGDR, and not the restarts.

    Args:
        optimizer (Optimizer): Wrapped optimizer.
        T_max (int): Maximum number of iterations.
        eta_min (float): Minimum learning rate. Default: 0.
        last_epoch (int): The index of last epoch. Default: -1.

    .. _SGDR\: Stochastic Gradient Descent with Warm Restarts:
        https://arxiv.org/abs/1608.03983
    """

    def __init__(self, optimizer, T_max, eta_min=0, last_epoch=-1):
        self.T_max = T_max
        self.eta_min = eta_min
        super(LegacyCosineAnnealingLR, self).__init__(optimizer, last_epoch)

    def get_lr(self):
        return [self.eta_min + (base_lr - self.eta_min) *
                (1 + math.cos(math.pi * self.last_epoch / self.T_max)) / 2
                for base_lr in self.base_lrs]


model = nn.Linear(10, 2)
optimizer = optim.SGD(model.parameters(), lr=1.)
steps = 10
scheduler = LegacyCosineAnnealingLR(optimizer, steps)

for epoch in range(2):
    for idx in range(steps):
        scheduler.step()
        print(scheduler.get_lr())
        print(optimizer.param_groups[0]['lr'])

> [1.0]
1.0
[0.9755282581475768]
0.9755282581475768
[0.9045084971874737]
0.9045084971874737
[0.7938926261462366]
0.7938926261462366
[0.6545084971874737]
0.6545084971874737
[0.5]
0.5
[0.34549150281252633]
0.34549150281252633
[0.2061073738537635]
0.2061073738537635
[0.09549150281252633]
0.09549150281252633
[0.024471741852423234]
0.024471741852423234
[0.0]
0.0
[0.024471741852423123]
0.024471741852423123
[0.09549150281252622]
0.09549150281252622
[0.20610737385376338]
0.20610737385376338
[0.3454915028125262]
0.3454915028125262
[0.4999999999999999]
0.4999999999999999
[0.6545084971874736]
0.6545084971874736
[0.7938926261462365]
0.7938926261462365
[0.9045084971874737]
0.9045084971874737
[0.9755282581475768]
0.9755282581475768
```

## Environment

 - PyTorch Version (e.g., 1.0): `1.0.0.dev20190312`
 - OS (e.g., Linux): Ubuntu 18.04
 - How you installed PyTorch (`conda`, `pip`, source): conda nightlies
 - Python version: 3.7

## Additional context

This issue might to be related to the refactoring of all schedulers done [here](https://github.com/pytorch/pytorch/pull/14010).

CC @chandlerzuo @ezyang

### PR DESCRIPTION
**Pull request by chandlerzuo**
Added the formula for the corner case. Updated unit tests.

Fixes #17913

### CODE REVIEW

**Comment by chandlerzuo** [comment]
If this is approved, we can abandon https://github.com/pytorch/pytorch/pull/19132

**Comment by ezyang** [comment]
I didn't do an in-depth math check (yet), but the preliminary shape looks good to me. cc @ptrblck @Kindpire

**Comment by chandlerzuo** [comment]
One thing I should mention is that, there are multiple induction formulas we can use for cosine annealing LR while still maintaining BC. Basically, the learning rate can be decomposed as constant + cyclic function of t. We have different choices for the constant here; it can be \eta_min, \eta_max, (\eta_min+\eta_max), or something else. Depending on the choice for the constant, we can work out the corresponding induction formula based on the remaining cyclic function.

Solution here treats \eta_min as the constant. I am not sure if it is the best one, but I think it is reasonable.

File: torch/optim/lr_scheduler.py [hunk 1]
```diff
@@ -226,7 +226,11 @@ class CosineAnnealingLR(_LRScheduler):
 
     .. math::
         \eta_{t+1} = \eta_{min} + (\eta_t - \eta_{min})\frac{1 +
-        \cos(\frac{T_{cur+1}}{T_{max}}\pi)}{1 + \cos(\frac{T_{cur}}{T_{max}}\pi)}
+        \cos(\frac{T_{cur+1}}{T_{max}}\pi)}{1 + \cos(\frac{T_{cur}}{T_{max}}\pi)},
+        T_{cur} \neq (2k+1)T_{max};\\
```
Comment (ezyang) [inline torch/optim/lr_scheduler.py]: Verified this condition. It's so confusing that we use a different `cur+1` convention in the code and math lol.

File: torch/optim/lr_scheduler.py [hunk 2]
```diff
@@ -259,6 +263,11 @@ def __init__(self, optimizer, T_max, eta_min=0, last_epoch=-1):
     def get_lr(self):
         if self.last_epoch == 0:
             return self.base_lrs
+        elif (self.last_epoch - 1 - self.T_max) % (2 * self.T_max) == 0:
+            return [group['lr'] + (base_lr - self.eta_min) *
+                    (1 - math.cos(math.pi / self.T_max)) / 2
+                    for base_lr, group in
+                    zip(self.base_lrs, self.optimizer.param_groups)]
```
Comment (ezyang) [inline torch/optim/lr_scheduler.py]: I spent a while staring at this for a while and couldn't figure out if the equation was right. It must be right, since the tests pass. Would you mind saying a little more about the derivation here?

Comment (chandlerzuo) [reply torch/optim/lr_scheduler.py]: latex_4a12231ae15a2d9abc1c42fea250a362

Comment (ezyang) [reply torch/optim/lr_scheduler.py]: Awesome, clears it up!

**Review by ezyang — APPROVED** [review]
